Readme.md

# cluster-mail

  Email notification plugin for Cluster

## Installation

    $ npm install cluster-mail

## Usage

 Initialize mail plugin with the given `email(s)`, with the given `options`.

 Options:

  - `from` sender email
  - `timeout` sendmail timeout in milliseconds
  - `subject` defaulting to "cluster({worker}) exception: {message}"
  - `template` function called with local variables (usually jade / ejs template etc)

## Example

    var http = require('http')
      , cluster = require('cluster')
      , mail = require('cluster-mail');

    var server = http.createServer(function(req, res){
      if (Math.random() > 0.9) throw new Error('fail!');
      res.end('Hello World');
    });

    cluster = cluster(server)
      .use(mail('your@email.com', { from: 'me@somewhere.com' }))
      .use(mail(['or@a-bunch.com', 'of@emails-here.com']))
      .listen(3000);

 By default cluster-mail can only notify uncaughtExceptions, however it is typically a good idea to do so from within your application's error handler as well. Below we invoke `cluster.mailException()`, where `cluster` is the return value of `cluster(server)`. We can pass an object containing data to add to the report, such as the authenticated user's id, email, request data, etc.
 
     app.error(function(err, req, res, next){
       cluster.mailException(err, {
           method: req.method
         , url: req.url
         , headers: req.headers
       });
     });

## Manual Reporting Example

  Often it is useful to report an exception within a worker that was caught, thus never terminating the process, in turn never notifying cluster's master process. An example of this would be within a connect or express error handler, where you simply responded with an error page, however you still want to be notified via mail.

  To do this we invoke the `mailException()` method with the error, as well as an optional object containing additional data to display in the email. This may include request information, user information etc. 

      // somewhere in your application
      var data = { method: req.method, url: req.url };
      cluster.mailException(err, data);

      // start the server
      cluster = cluster(server)
        .use(mail(['dev@learnboost.com', 'errors@learnboost.com']))
        .listen(3000);

## Screenshot

![](http://f.cl.ly/items/0K0F3t1s2o172b0j2407/Screenshot.png)

## License 

(The MIT License)

Copyright (c) 2011 LearnBoost &lt;tj@learnboost.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

0.1.4 / 2011-10-06
==================

  * fixed; pass along the worker so id displays in subject line

0.1.3 / 2011-10-03
==================

  * fixed; stopped sending same subject line for every email

0.1.2 / 2011-09-26
==================

    * Fixed; calling mailException from worker before worker recieves connect event

0.1.1 / 2011-05-16 
==================

  * Fixed bug causing several of the same subject

0.1.0 / 2011-04-19 
==================

  * Added arbitrary details display (request url, method, headers, etc)
  * Added `cluster.mailException()` for workers
  * Added `send()` helper function

0.0.2 / 2011-04-05 
==================

  * Fixed; output stack / error message

0.0.1 / 2010-01-03
==================

  * Initial release


lib/cluster-mail.js

/*!
 * cluster-mail
 * Copyright(c) 2011 LearnBoost <tj@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var Email = require('email').Email
  , jade = require('jade')
  , fs = require('fs');

/**
 * Expose the plugin.
 */

exports = module.exports = mail;

/**
 * Library version.
 */

exports.version = '0.1.4';

/**
 * Initialize mail plugin with the given `email(s)`,
 * with the given `options`.
 *
 * Options:
 *
 *    - `from` sender email
 *    - `timeout` sendmail timeout in milliseconds
 *    - `subject` defaulting to " cluster({worker}) exception: {message}"
 *    - `template` function called with local variables (usually jade / ejs template etc)
 *
 * @param {String|Array} email(s)
 * @param {Object} options
 * @return {Function}
 * @api public
 */

function mail(emails, options) {
  // options
  options = options || {};
  emails = emails || [];
  if ('string' == typeof emails) emails = [emails];
  if (!emails.length) throw new Error('email(s) required');

  // defaults
  options.to = emails;
  options.from = options.from || 'mail@cluster.com';
  options.subject = options.subject || 'cluster({worker}) exception: {message}';
  options.template = options.template || template();

  mail.enableInWorker = true;

  function mail(master){
    if (master.isWorker) {
      master.mailException = function(err, data){
        send(this.worker, err, options, data);
      };
    } else {
      master.mailException = function(worker, err, data){
        send(worker, err, options, data);
      };

      master.on('worker exception', function(worker, err){
        send(worker, err, options);
      });
    }

  };

  return mail;
};

/**
 * Send `err` notification for the given `worker`.
 *
 * @param {Worker} worker
 * @param {Error} err
 * @param {Object} options
 * @param {Object} data
 * @api private
 */

function send(worker, err, options, data) {
  options = clone(options);

  // subject
  options.subject = options.subject
    .replace('{worker}', worker.id)
    .replace('{message}', err.message);

  // body
  options.bodyType = 'html';
  options.body = options.template({
      worker: worker
    , error: err
    , data: data
  });

  // send 
  var email = new Email(options);
  email.send(function(err){
    if (err) console.error(err.stack || err.message);
  });
}

/**
 * Clone the given `obj`.
 *
 * @param {Object} obj
 * @return {Object}
 * @api private
 */

function clone(obj){
  var ret = {};
  for (var key in obj) ret[key] = obj[key];
  return ret;
}

/**
 * Read / compile default template.
 */

function template() {
  var str = fs.readFileSync(__dirname + '/email.jade', 'utf8');
  return jade.compile(str);
}


example.js

/**
 * Module dependencies.
 */

var http = require('http')
  , cluster = require('cluster')
  , mail = require('./');

var server = http.createServer(function(req, res){
  // uncaught exceptions are automatically reported, however
  // for exceptions that we have caught, typically within
  // an error handler, we may report them with cluster.reportException()
  try {
    if (Math.random() > 0.9) throw new Error('fail!');
    res.end('Hello World');
  } catch (err) {
    // we can also optionally send an arbitrary json object
    // for use within our template
    var data = { method: req.method, url: req.url, headers: req.headers };
    cluster.mailException(err, data);
    console.log('sent %s', err.message);
  }
});

cluster = cluster(server)
  .use(mail('tjholowayhuk@gmail.com'))
  .listen(3000);